docs/design/ha-flow/ha-flow-nb-api.md

```diff
@@ -0,0 +1,406 @@
+}
+```
+
+`sub_flows` contain info about A-end and B-end endpoints. It can contain 2 records.
+
+`status` calculated same way as common flow.
+It will `UP` if all segments of HA-flow are `UP`. 
+
+It will be `DEGRADED` in one of following cases:
+1. Primary path is `UP`, but protected path is `DOWN`/`DEGRADED`.
+2. * HA-flow has path computation strategy 'MAX_LATENCY'
+   * and primary or protected path doesn't fit into max_latency requirements
+   * but fit into non-zero max_latency_tier2 requirements
+
+It will be `DOWN` in one of following cases:
+1. Any path segment is `DOWN`
+2. * HA-flow has path computation strategy 'MAX_LATENCY'
+   * and primary or protected path doesn't fit into max_latency requirements
+   * and max_latency_tier2 is equal to 0/null.
+3. * HA-flow has path computation strategy 'MAX_LATENCY'
+   * and primary or protected path doesn't fit into max_latency_tier2 requirements.
+
+## Update operation
+
+REST endpoints: 
+ * `PUT /v2/ha-flows/{ha_flow_id}`
+ * `PATCH /v2/ha-flows/{ha_flow_id}`
+
+Request/response bodies are the same to the create operation. In case of PATCH request
+only present in request body fields will be updated.
+
+## Reroute operation
+
+REST endpoint: `POST /v2/ha-flows/{ha_flow_id}/reroute`
+
+Request body is ignored, response payload:
+
+```json
+{
+  "rerouted": true,
+  "shared_path": {
+    "nodes": [
+      {
+        "port_no": 0,
+        "segment_latency": 10,
+        "switch_id": "00:00:00:00:00:00:00:04"
+      }
+    ]
+  },
+  "sub_flow_paths": [
+    {
+      "flow_id": "fAAAAAAAAAAAAAAA",
+      "nodes": [
+        {
+          "port_no": 0,
+          "segment_latency": 0,
+          "switch_id": "00:00:00:00:00:00:00:05"
+        }
+      ]
+    },
+    {
+      "flow_id": "fAAAAAAAAAAAAAAB",
+      "nodes": [
+        {
+          "port_no": 0,
+          "segment_latency": 0,
+          "switch_id": "00:00:00:00:00:00:00:06"
+        }
+      ]
+    }
+  ]
+}
+```
+
+## Delete operation
+
+REST endpoint: `DELETE /v2/ha-flows/{ha_flow_id}`
+
+Request body is ignored, response body is the same to the read operation.
+
+## Read paths
+
+REST endpoint: `GET /v2/ha-flows/{ha_flow_id}/paths`
+
+Response body:
+
+```json
+{
+  "shared_path": {
+    "nodes": [
+      {
+        "port_no": 0,
+        "segment_latency": 0,
+        "switch_id": "00:00:00:00:00:00:00:08"
+      }
+    ]
+  },
+  "sub_flow_paths": [
+    {
+      "flow_id": "fAAAAAAAAAAAAAAA",
+      "nodes": [
+        {
+          "port_no": 0,
+          "segment_latency": 0,
+          "switch_id": "00:00:00:00:00:00:00:09"
+        }
+      ]
+    },
+    {
+      "flow_id": "fAAAAAAAAAAAAAAB",
+      "nodes": [
+        {
+          "port_no": 0,
+          "segment_latency": 0,
+          "switch_id": "00:00:00:00:00:00:00:09"
+        }
+      ]
+    }
+  ]
+}
+```
+
+## Ping operation
+
+REST endpoint: `POST /v2/ha-flows/{ha_flow_id}/ping`
+
+request payload:
+
+```json
+{
+  "timeout": 0
```

Comment:
Just curious, it is going to send one ICMP, but receive 2 replies? What if their timeouts are different? Is it enough just send one ping?

Reply:
It's not an ICMP. In Kilda Ping packet is custom UDP. Looks like this https://github.com/telstra/open-kilda/blob/2eb4f45072a063421299ecc8c6c099b8c8ff70f0/src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/service/ping/PingService.java#L91

How looks like ping process for common flow:
Kilda generates 2 UDP packets and send one from src endpoint - one from dst. After that Kilda expect to catch them at dst and src. 

How it will look like in HA-flow case:

Kilda will generate 3 packets for Z,A,B-ends.
After that Kilda will expect to catch:

2 packets at Z end.
1 packet at A-end
1 packet at B-end

Reply:
aha, got it. Thanks for the explanation! 👍